src/common/cockpitwebserver.c

```diff
@@ -1028,46 +1028,60 @@ on_request_input (GObject *pollable_input,
   gsize length;
   gssize count;
 
-  length = request->buffer->len;
-  g_byte_array_set_size (request->buffer, length + 4096);
-
-  count = g_pollable_input_stream_read_nonblocking (input, request->buffer->data + length,
-                                                    4096, NULL, &error);
-  if (count < 0)
+  /* read all pending data */
+  for (;;)
     {
-      g_byte_array_set_size (request->buffer, length);
+      length = request->buffer->len;
+      g_byte_array_set_size (request->buffer, length + 4096);
 
-      /* Just wait and try again */
-      if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_WOULD_BLOCK))
+      count = g_pollable_input_stream_read_nonblocking (input, request->buffer->data + length,
+                                                        4096, NULL, &error);
+      if (count < 0)
         {
+          g_byte_array_set_size (request->buffer, length);
+
+          if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_WOULD_BLOCK))
+            {
+              /* finished reading everything that was pending */
+              g_error_free (error);
+              break;
+            }
+
+          if (!should_suppress_request_error (error, length))
+            g_message ("couldn't read from connection: %s", error->message);
+
+          cockpit_request_finish (request);
           g_error_free (error);
-          return TRUE;
+          return FALSE;
         }
 
-      if (!should_suppress_request_error (error, length))
-        g_message ("couldn't read from connection: %s", error->message);
+      g_byte_array_set_size (request->buffer, length + count);
 
-      cockpit_request_finish (request);
-      g_error_free (error);
-      return FALSE;
-    }
-
-  g_byte_array_set_size (request->buffer, length + count);
+      if (count == 0)
+        {
+          if (request->eof_okay)
+            close_io_stream (request->io);
+          else
+            g_debug ("caller closed connection early");
+          if (request->buffer->len == 0)
+            {
+              cockpit_request_finish (request);
+              return FALSE;
+            }
+          else
+            {
+              break;
+            }
+        }
 
-  if (count == 0)
-    {
-      if (request->eof_okay)
-        close_io_stream (request->io);
-      else
-        g_debug ("caller closed connection early");
-      cockpit_request_finish (request);
-      return FALSE;
+      /* Once we receive data EOF is unexpected (until possible next request) */
+      request->eof_okay = FALSE;
     }
 
-  /* Once we receive data EOF is unexpected (until possible next request) */
-  request->eof_okay = FALSE;
-
-  return parse_and_process_request (request);
+  if (request->buffer->len > 0)
+    return parse_and_process_request (request);
```

Comment:
I think that this should be called within the loop for each successful read. It will both simplify the code, and also remain true to the original intent of how that function affects the request structure.

Reply:
Interesting, I explicitly took it out of the loop, as it seems futile -- if there is more data, then the request definitively isn't complete, and `parse_and_process_request()` would just return False (hopefully), asking for more data.

Reply:
It seems that parse_and_process_request() could update request ... thereby affecting the loop. I think it's better just to call it iteratively.

Reply:
Oh hmmm. in addition there's another bug here. We should be calling parse_and_process_request until it returns FALSE. 

If indeed edge triggers, we not only need to read all input in a loop, but we need to process all data in the buffer in a loop.

Reply:
I'm afraid I still don't understand what you mean. `parse_and_process_request` sets `again` to true exactly when there is not enough data yet in the buffer, i. e. usually there's no terminating `\r\n` for the headers yet. This won't ever change when just calling it several times.

I initially thought that it will also be useless if we already know that there is more data in the buffer, as the request is clearly incomplete -- but apparently I misunderstand this, so I'm ok with moving it into the loop. But I'm not sure what you mean by "calling it until it returns FALSE"? If there's no more data in the buffer, and it returns FALSE, then we just leave the callback and wait for more data -- i. e. exactly how it behaves in current master. This PR doesn't change the request processing logic, just flushing the input buffer in the callback.

Reply:
Oh, do you mean that the `CockpitRequest` buffer could contain more than one request?

Reply:
I stared at this for a while, and I'm afraid I'm unable to see how `parse_and_process_request()` can sensibly be moved into the loop. If there is more data in the buffer and it returns FALSE, we have a contradiction -- it clearly processed the request too early and didn't notice that it was incomplete. Also, if it returns TRUE and we have no more data in the buffer, we must break the loop (and the current PR covers this case already).

The "good" cases are that we have no more data in the buffer and it returns FALSE (and the current PR covers that well), or that we have more data in the buffer and it returns TRUE (and the PR avoids this case, as that seems to be wasted time). So I'd appreciate your help with moving it into the loop.

Reply:
To clarify: The intention of this PR is to essentially change the g_pollable_input_stream_read_nonblocking() from a fixed buffer size to "as much as you have pending", not more, not less.